cpp/include/rapidsmpf/shuffler/chunk_batch.hpp

```diff
@@ -0,0 +1,390 @@
+        return header()->num_chunks;
+    }
+
+    /**
+     * @brief Access the id of the chunk batch.
+     * @return The id of the chunk batch.
+     */
+    [[nodiscard]] uint32_t id() const {
+        return header()->id;
+    }
+
+    /**
+     * @brief Releases the ownership of the metadata buffer of the chunk batch.
+     *
+     * @return The released metadata buffer.
+     *
+     * @throws std::logic_error if the metadata buffer is null.
+     *
+     * @note This method should be called only once.
+     */
+    [[nodiscard]] std::unique_ptr<std::vector<uint8_t>> release_metadata() {
+        RAPIDSMPF_EXPECTS(metadata_buffer_, "metadata buffer is null");
+        return std::move(metadata_buffer_);
+    }
+
+    /**
+     * @brief Releases the ownership of the payload buffer of the chunk batch.
+     *
+     * @return The released payload buffer. If there was no payload, returns nullptr.
+     *
+     * @note Unlike metadata buffer, payload buffer could be null.
+     */
+    [[nodiscard]] std::unique_ptr<Buffer> release_payload() {
+        return std::move(payload_data_);
+    }
+
+    /**
+     * @brief Creates a chunk batch.
+     *
+     * @param id The id of the batch.
+     * @param dest_rank The destination rank of the batch.
+     * @param chunks The chunks to be included in the batch.
+     * @param br The buffer resource to use for allocating the metadata buffer.
+     * @param stream The stream to use for allocating the metadata buffer.
+     * @return New chunk batch.
+     *
+     * @throws std::logic_error if the payload buffer memory types of the chunks are not
+     * the same. Chunk data is expected to be of the same memory type.
+     */
+    [[nodiscard]] static ChunkBatch create(
+        uint32_t id,
+        Rank dest_rank,
+        std::vector<Chunk>&& chunks,
+        BufferResource* br,
+        rmm::cuda_stream_view stream
+    );
+
+    /**
+     * @brief Creates a chunk batch from a metadata buffer and a payload buffer.
+     *
+     * @param metadata The metadata buffer.
+     * @param payload_data The payload buffer.
+     * @return New chunk batch.
+     *
+     * @throws std::logic_error if the provided buffers violate the format of a chunk
+     * batch.
+     */
+    [[nodiscard]] static ChunkBatch create(
+        std::unique_ptr<std::vector<uint8_t>> metadata,
+        std::unique_ptr<Buffer> payload_data
+    );
+
+    /**
+     * @brief Visits the chunk data in the batch.
+     *
+     * @tparam VisitorFn Visitor function type. Must be callable with the following
+     * signature:
+     * void(Chunk::MetadataMessageHeader const* chunk_header,
+     *      std::vector<uint8_t> const& metadata_buf,
+     *      std::ptrdiff_t metadata_offset,
+     *      Buffer const& payload_buf,
+     *      std::ptrdiff_t payload_offset)
+     * @param visitor visitor function
+     */
+    template <typename VisitorFn>
+    void visit_chunk_data(VisitorFn visitor) const {
```

Comment:
I think I don't really like this interface and would prefer to only have the iterator-like access. WDYT?